unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts

```diff
@@ -75,413 +61,477 @@ export abstract class DataParser {
+      | string
+      | {
+          flags?: string;
+          encoding?: BufferEncoding;
+          fd?: number;
+          mode?: number;
+          autoClose?: boolean;
+          emitClose?: boolean;
+          start?: number;
+          end?: number;
+          highWaterMark?: number;
+        }
+  ) {
+    this.path = path;
+    this.options = options;
+
+    this.isParsed = false;
+    this.isBuilt = false;
+    this.isTraversed = false;
+
+    this.data = [];
+    this.graph = {};
+    this.cycles = [];
+  }
+
+  public abstract parseData(): Promise<IData>;
+  public abstract buildGraph(): IGraph;
+  public abstract detectCycles(): INode[];
+  public abstract isValid(): boolean;
+
+  public getData() {
+    return this.data;
+  }
+  public getGraph() {
+    return this.graph;
+  }
 }
 
 export class CSVParser extends DataParser {
-    constructor(path: fs.PathLike, options?: string | {
-        flags?: string;
-        encoding?: BufferEncoding;
-        fd?: number;
-        mode?: number;
-        autoClose?: boolean;
-        emitClose?: boolean;
-        start?: number;
-        end?: number;
-        highWaterMark?: number;
-    }) {
-        super(path, options);
+  constructor(
+    path: fs.PathLike,
+    options?:
+      | string
+      | {
+          flags?: string;
+          encoding?: BufferEncoding;
+          fd?: number;
+          mode?: number;
+          autoClose?: boolean;
+          emitClose?: boolean;
+          start?: number;
+          end?: number;
+          highWaterMark?: number;
+        }
+  ) {
+    super(path, options);
+  }
+
+  public detectCycles(): INode[] {
+    if (this.isTraversed) return this.cycles;
+
+    const visited = {};
+    const stack = [this.graph.root];
+    while (stack.length > 0) {
+      const entity = stack.pop();
+
+      if (!visited[entity.code]) {
+        visited[entity.code] = true;
+      }
+
+      if (entity.children) {
+        for (let i = 0; i < entity.children.length; i++) {
+          const child = entity.children[i];
+          if (!visited[child.code]) {
+            stack.push(child);
+          } else {
+            this.cycles.push(child);
+          }
+        }
+      }
     }
 
-    public detectCycles(): INode[] {
-        if (this.isTraversed) return this.cycles;
+    return this.cycles;
+  }
+
+  public isValid(): boolean {
+    return !!this.detectCycles();
+  }
+
+  public async parseData(): Promise<IData> {
+    if (this.isParsed) return this.data;
+
+    // Read data stream.
+    const stream = await fs.createReadStream(this.path, this.options);
+    await new Promise((resolve) => {
+      stream
+        .pipe(csv())
+        .on('data', (row) => {
+          const entity = Object.entries<string>(row).reduce(
+            (acc: IRow, [column, value]: [string, string]) => {
+              const key = column
+                .trim()
+                .toLowerCase()
+                .replace(/ *\([^)]*\) */g, '')
+                .replace(/ /g, '_');
```

Comment:
Perhaps we could define these regex with a constant? Or just a comment would be fine, too!

Reply:
I like constants. Though I also had to comment the really dodgy one.
61e859237